rest/data/project.go

```diff
@@ -80,6 +80,81 @@ func (pc *DBProjectConnector) UpdateRepo(repoRef *model.RepoRef) error {
+func (pc *DBProjectConnector) SaveProjectSettingsForSection(ctx context.Context, projectId string, changes *restModel.APIProjectSettings,
+	section model.ProjectRefSection, userId string) error {
+	before, err := model.GetProjectSettingsById(projectId)
+	if err != nil {
+		return errors.Wrap(err, "error getting before project settings event")
+	}
+	v, err := changes.ProjectRef.ToService()
+	if err != nil {
+		return errors.Wrap(err, "error converting project ref")
+	}
+	newProjectRef := v.(*model.ProjectRef)
+	// if the project ref doesn't use the repo, then this will just be the same as newProjectRef
+	// used to verify that if something is set to nil in the request, we properly validate using the merged project ref
+	mergedProjectRef, err := model.GetProjectRefMergedWithRepo(*newProjectRef)
+	if err != nil {
+		return errors.Wrapf(err, "error merging project ref")
+	}
+
+	catcher := grip.NewBasicCatcher()
+	modified := false
+	switch section {
+	// aliases and notifications use a different connector and should be handled separated
+	case model.ProjectRefGeneralSection:
+		// check if webhook is enabled if the owner/repo has changed
+		if mergedProjectRef.Owner != before.ProjectRef.Owner || mergedProjectRef.Repo != before.ProjectRef.Repo {
+			_, err = pc.EnableWebhooks(ctx, mergedProjectRef)
+			if err != nil {
+				return errors.Wrapf(err, "Error enabling webhooks for project '%s'", projectId)
+			}
+		}
+	case model.ProjectRefAccessSection:
+		// For any admins that are only in the original settings, remove access.
+		// For any admins that are only in the updated settings, give them access.
+		adminsToDelete, adminsToAdd := utility.StringSliceSymmetricDifference(before.ProjectRef.Admins, mergedProjectRef.Admins)
+		if err = pc.UpdateAdminRoles(newProjectRef, adminsToAdd, adminsToDelete); err != nil {
+			return errors.Wrap(err, "error updating admin roles")
+		}
+		modified = true
+		if !before.ProjectRef.IsRestricted() && mergedProjectRef.IsRestricted() {
+			catcher.Wrap(before.ProjectRef.MakeRestricted(), "error making project restricted")
+		}
+		if before.ProjectRef.IsRestricted() && !mergedProjectRef.IsRestricted() {
+			catcher.Wrap(before.ProjectRef.MakeUnrestricted(), "error making unrestricted")
+		}
+
+	case model.ProjectRefVariablesSection:
+		// Remove any variables that only exist in the original settings. (It's possible we can send this information
+		// from the UI in the future instead)
+		toDelete := []string{}
+		for key, _ := range before.Vars.Vars {
+			if _, ok := changes.Vars.Vars[key]; !ok {
+				toDelete = append(toDelete, key)
+			}
+		}
+		changes.Vars.VarsToDelete = toDelete
+		if err = pc.UpdateProjectVars(projectId, &changes.Vars, false); err != nil { // destructively modifies vars
+			return errors.Wrapf(err, "Database error updating variables for project '%s'", projectId)
+		}
+	}
+	// save project ref only after the rest of the section has been validated.
```

Comment:
what do you mean by this? Where is it validated?

Reply:
I think this is a future note; we discussed validating in the resolver

Reply:
Rearranged for clarity